src/odin/ingestion/spare/spare_job.py

```diff
@@ -0,0 +1,400 @@
+    # metadata field is arbitrary dict, so reencode back to json string
+    if "metadata" in columns:
+        for row in j:
+            row["metadata"] = json.dumps(row["metadata"])
+    df = pl.DataFrame(j, columns)
+    return df
+
+
+def load_parquet(parquet_path: str, columns: Columns) -> pl.DataFrame:
+    """
+    Get a parquet file and return it as a DataFrame.
+
+    Path may start with s3:// or be a local filesystem file.
+    If the file doesn't exist, return an empty DataFrame with the columns from the schema.
+    """
+    columns = METADATA_COLUMNS | columns
+    # TODO assert that loaded file has same columns in same order
+    # TODO save the file timestamp to catch conflicts
+    if parquet_path.startswith("s3://"):
+        if s3.object_exists(parquet_path):
+            return pl.read_parquet(parquet_path)
+        else:
+            return pl.DataFrame([], columns)
+    else:
+        try:
+            return pl.read_parquet(parquet_path)
+        except FileNotFoundError:
+            return pl.DataFrame([], columns)
+
+
+def write_parquet(df: pl.DataFrame, parquet_path: str) -> None:
+    """
+    Write a dataframe to a parquet file.
+
+    Path may start with s3:// or be a local filesystem file.
+
+    Logically, we only append to the parquet file,
+    but appending row groups is not well supported and is overkill for this use,
+    so we write the whole file.
+
+    TODO Prevent overwriting data if there's a conflict:
+    It's possible that the output parquet file was modified (perhaps by another instance of Odin)
+    after reading the file, while crunching the data. If so, we should not write the output file,
+    as that would overwrite those modifications
+    Instead, abort the job, leave the input files unarchived, and re-run the job immediately.
+    This way, the input files can be reprocessed and changes calculated based on the newest version
+    of the parquet file.
+    Implementation notes: For AWS, Save the ETag on read and pass it to If-Match on write.
+    For local files, use os.path.getmtime.
+    """
+    if parquet_path.startswith("s3://"):
+        df.write_parquet(parquet_path)
+    else:
+        os.makedirs(os.path.dirname(parquet_path), exist_ok=True)
+        df.write_parquet(parquet_path)
+
+
+def datetime_from_delta_path_ms(path: str) -> int:
+    """
+    Extract a datetime from the filenames generated by Delta
+
+    Returns number of ms since the epoch.
+    Delta makes filenames like
+    2025/04/01/2025-04-01T18:12:33Z_https_example.com_path.gz
+    This extracts extracts and parses the ISO8601 timestamp
+    It ignores the yyyy/mm/dd/ prefix so it works if the file has been copied elsewhere.
+    """
+    iso8601 = path.split("/")[-1].split("_")[0]
+    if not iso8601.endswith("Z"):
+        raise Exception(
+            f'Expected iso8601 timestamp with a Z. Got "{iso8601}". From path "{path}".'
+        )
+    return int(isoparse(iso8601).timestamp() * 1000)
+
+
+def merge_new_data(
```

Comment:
Suggestion: are the `created_at` and `updated_at` fields available in this dataset reliable? If so, those fields, in combination with an `id` field, could simplify this process.